internal/cli/serve.go

```diff
@@ -81,3 +106,75 @@ func runService(stopCh chan os.Signal, service http.Server, airlock server.Airlo
 	}
 	stopCh <- os.Interrupt
 }
+
+// checkConsistency continuously checks for consistency between configuration and remote state.
+//
+// It takes care of polling etcd, exposing the shared state as metrics, and warning if
+// it detects a mismatch with the service configuration.
+func checkConsistency(ctx context.Context, service server.Airlock) {
+	prometheus.MustRegister(configGroups)
+	prometheus.MustRegister(configSlots)
+	prometheus.MustRegister(databaseLocks)
+	prometheus.MustRegister(databaseSlots)
+
+	configGroups.Set(float64(len(service.LockGroups)))
+	for group, maxSlots := range service.LockGroups {
+		configSlots.WithLabelValues(group).Set(float64(maxSlots))
+	}
+
+	// Consistency-checking logic, with its own scope for defers.
+	checkAndLog := func() {
+		for group, maxSlots := range service.LockGroups {
+			innerCtx, cancel := context.WithTimeout(ctx, service.EtcdTxnTimeout)
+			defer cancel()
+
+			// TODO(lucab): re-arrange so that the manager can be re-used.
+			manager, err := lock.NewManager(innerCtx, service.EtcdEndpoints, group, maxSlots)
+			if err != nil {
+				logrus.WithFields(logrus.Fields{
+					"reason": err.Error(),
+				}).Warn("consistency check, manager creation failed")
+				continue
+			}
+			semaphore, err := manager.FetchSemaphore(innerCtx)
+			if err != nil {
+				logrus.WithFields(logrus.Fields{
+					"reason": err.Error(),
+				}).Warn("consistency check, semaphore fetch failed")
+				continue
+			}
+
+			// Update metrics.
+			databaseSlots.WithLabelValues(group).Set(float64(semaphore.TotalSlots))
+			databaseLocks.WithLabelValues(group).Set(float64(len(semaphore.Holders)))
+
+			// Log any inconsistencies.
+			if semaphore.TotalSlots != maxSlots {
+				logrus.WithFields(logrus.Fields{
+					"config":   maxSlots,
+					"database": semaphore.TotalSlots,
+					"group":    group,
+				}).Warn("semaphore max slots consistency check failed")
+			}
+			if semaphore.TotalSlots < uint64(len(semaphore.Holders)) {
+				logrus.WithFields(logrus.Fields{
+					"group":  group,
+					"holder": len(semaphore.Holders),
+					"slots":  semaphore.TotalSlots,
+				}).Warn("semaphore locks consistency check failed")
+			}
+		}
+	}
+
+	for {
+		checkAndLog()
+
+		pause := time.NewTimer(time.Minute)
```

Comment:
Any thought on using a [`Ticker`](https://golang.org/pkg/time/#NewTicker) rather than a `Timer` to avoid the re-instantiation of an object every loop?

Reply:
Oh, definitely! I somehow missed it. I'll rework that part tomorrow.

Reply:
After a second thought, I think I prefer the status quo.
The current behavior results in equal-width 60s pauses. Changing to a Ticker would result in dynamic pause of 60s minus the duration of `checkAndLog()`.